service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,209 @@
+                    ErrorAttribute.INVALID_GIFT_CERTIFICATE_ID_ERROR, id);
+        }
+    }
+
+    @Override
+    public boolean update(String id, GiftCertificate newCertificate) {
+        try {
+            GiftCertificate oldCertificate = dao.findById(Long.parseLong(id)).orElseThrow(() ->
+                    new ResourceNotFoundException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                            ErrorAttribute.RESOURCE_NOT_FOUND_ERROR, id));
+
+            if (updateCertificateFields(oldCertificate, newCertificate)) {
+                oldCertificate.setLastUpdateDate(LocalDateTime.now());
+                if (!CollectionUtils.isEmpty(oldCertificate.getTags())) {
+                    List<Tag> newTags = (List<Tag>) CollectionUtils.removeAll(oldCertificate.getTags(),
+                            CollectionUtils.intersection(tagService.findAll(0, 0), oldCertificate.getTags()));
+                    newTags.forEach(tagService::insert);
+                }
+                dao.update(oldCertificate);
+            } else {
+                throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                        ErrorAttribute.INVALID_GIFT_CERTIFICATE_ERROR, newCertificate.toString());
+            }
+        } catch (NumberFormatException e) {
+            throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                    ErrorAttribute.INVALID_GIFT_CERTIFICATE_ID_ERROR, id);
+        }
+        return true;
+    }
+
+    @Override
+    public void disconnectTagById(String tagId) {
+        Tag tag = tagService.findById(tagId);
+        List<GiftCertificate> certificatesWithSuchTag = findCertificatesWithTagsByCriteria(0, 0,
+                Collections.singletonList(tag.getName()), null, null, null, null);
+        if (!CollectionUtils.isEmpty(certificatesWithSuchTag)) {
+            for (GiftCertificate certificate : certificatesWithSuchTag) {
+                Set<Tag> updatedTags = certificate.getTags();
+                updatedTags.remove(tag);
+                certificate.setTags(updatedTags);
+                dao.update(certificate);
+            }
+        }
+    }
+
+    @Override
+    public GiftCertificate findById(String id) {
+        try {
+            return dao.findById(Long.parseLong(id)).orElseThrow(() -> new ResourceNotFoundException(
+                    ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE, ErrorAttribute.RESOURCE_NOT_FOUND_ERROR, id));
+        } catch (NumberFormatException e) {
+            throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                    ErrorAttribute.INVALID_GIFT_CERTIFICATE_ID_ERROR, id);
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> findAll(int page, int elements) {
+        return dao.findAll(page, elements);
+    }
+
+    @Override
+    public List<GiftCertificate> findCertificatesWithTagsByCriteria(int page, int elements, List<String> tagsNames,
```

Comment:
you now the parameter in the input, you know the operation which should be performed, think a bit how the below if-s can be transformed. Not to correct, just discuss.